src/common/tracer.h

```diff
@@ -0,0 +1,66 @@
+#ifndef TRACER_H_
+#define TRACER_H_
+
+#define SIGNED_RIGHT_SHIFT_IS 1
+#define ARITHMETIC_RIGHT_SHIFT 1
+
+#include <iostream>
+#include <arpa/inet.h>
+#include <yaml-cpp/yaml.h>
+
+#include <jaegertracing/Tracer.h>
+
+#ifndef WITH_JAEGER
+  #define WITH_JAEGER
+#endif
+
+typedef std::unique_ptr<opentracing::Span> Span;
+
+class Jager_Tracer{
+  public:
+    Jager_Tracer(){}
+    ~Jager_Tracer(){
+      if(this->tracer == NULL)
+	return;
+      if(!this->isTracerClosed)
+        this->tracer->Close();
+        this->isTracerClosed=true;
+    }
+
+    void init_tracer(const char* tracerName,const char* filePath){
+        auto yaml = YAML::LoadFile(filePath);
+        auto configuration = jaegertracing::Config::parse(yaml);
+        this->isTracerClosed=false;
+        this->tracer = jaegertracing::Tracer::make(
+        tracerName,
+        configuration,
+        jaegertracing::logging::consoleLogger());
+        opentracing::Tracer::InitGlobal(
+        std::static_pointer_cast<opentracing::Tracer>(tracer));
+    }
+    inline void finish_tracer(){
+      if(this->tracer == NULL)
+	return;
+      if(!this->isTracerClosed){
+	 this->isTracerClosed=true;
+	 this->tracer->Close();
+       }
+    }
+    Span new_span(const char* spanName){
+      Span span=opentracing::Tracer::Global()->StartSpan(spanName);
+      return std::move(span);
+    }
+    Span child_span(const char* spanName,const Span& parentSpan){
+      Span span = opentracing::Tracer::Global()->StartSpan(spanName, {opentracing::ChildOf(&parentSpan->context())});
+      return std::move(span);
+    }
+    Span followup_span(const char *spanName, const Span& parentSpan){
+    Span span = opentracing::Tracer::Global()->StartSpan(spanName, {opentracing::FollowsFrom(&parentSpan->context())});
+    return std::move(span);
+  }
+private:
+  std::shared_ptr<opentracing::v2::Tracer> tracer = NULL;
```

Comment:
why do you need a shared pointer?
there is a single owner to the tracer

Reply:
just use an unique_ptr

Reply:
Jaeger requires the tracer variable to be a shared_ptr for initialization, I dont know why, but the library demands it.

Reply:
@ideepika the spans can be shared too I making them shared, to adapt new changes

Reply:
> @yuvalif spans are supposed to be unique while tracer can be shared by multiple services for collecting spans

this does not make it a shared pointer, though. as long as ownership of the pointer is unique, the pointer should be a unique_ptr and not shared_ptr.
I think that in the rgw, we can make sure that rgw main is the owner of the pointer (the only point where it is constructed and destroyed).

Reply:
> @ideepika the spans can be shared too I making them shared, to adapt new changes

i think that the rgw coroutine design, the span is not supposed to be passed between different threads. it should trickle through the call stack, until it could be ended. this means unique ownership.
if someone would like to change that design in the future, e.g. flows that branch into different threads - they should make the change into a shared pointer, but, for now, lets keep that a unique pointer

Reply:
@yuvalif I need to make the Span shared beacuse if do this 
```
class A{
Span a;
};
then to assign to `a` I cannot move pointer because  then the span will loose its owernship in which it was created, so to make this work I need to use `shared_ptr`

Reply:
**moving a pointer is not the same as moving ownership to the pointer**

you can move bare pointers as much as you want, as long as:
* you never use `delete`
* you know that the owner (whatever created and deletes them) is not going to leave them dangling for whatever uses them

for that you can have the rgw main (or some other common place) create the pointer as a `unique_ptr`
since you are juts passing a bare pointer to decedent function calls, this pointer will only be deleted when the `unique_ptr` goes out of scope in the original owner.
so, no double deletes, no leaks and no dangling pointers

Reply:
```
struct A{
Span span;
void op(){}
};

func(A){
Span span;
trace(span);
A abhi;
abhi.span = std::move(span); //looses ownership so this span wont be finished at end of this function but at the end of //abhi.op function
abhi.op();
}
```
A::op(){
trace(span);
}

Reply:
okay got this I will be using unique_ptr for Spans :)